Bail out of run-queue if another run is in progress.

barbun opened the pull request:
## Rationale
Consider a scenario where a site is using cron to run `quant:run-queue`, and the cron frequency is set to a fairly short time (e.g. 5 mins). Let's say there's also 1M nodes in the queue. With each new cron run the concurrency of jobs would pile up, potentially overwhelming the system.

## Proposed solution
Introduce a new option  `quant:run-queue --bail` that will check if another run is in progress, and if so, it will bail out of the current run.

Commit:
Added an optional bail out from seed-run if another run is in progress.

ivangrynenko commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,29 @@ private function getDrushPath() {
+   * @usage quant:run-queue --bail
    */
-  public function message($options = ['threads' => 5]) {
+  public function message($options = ['threads' => 5, 'bail' => false]) {
```
I recommend to have `'bail' => true` by default, otherwise we risk taking too much server CPU/RAM resources when command executed frequently, such as every 5 minutes via a cron job.

Commit:
Refactored implementation to lock queue run by default.

barbun commented:
Had a conversation with @ivangrynenko  and he made a good point that there's no reason why lock behaviour should not be the default.

PR has been refactored:
- Changed `--bail` option to `--unlock`
- Changed the code to use lock file by default, unless `--unlock` option was passed, which would allow users the option to still perform a run ignoring  the lock.

kepol commented:
Synced issue to drupal.org:

https://www.drupal.org/project/quantcdn/issues/3334336

kepol commented:
@ivangrynenko Are you able to review the changes and see if it works as expected?

kepol commented:
I'll review this.

kepol commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,29 @@ private function getDrushPath() {
    * @option threads
    *   Number of threads to use (default 5)
    * @usage quant:run-queue --threads=5
+   * @option unlock
+   * Allow a queue run even if another run is in progress. (default is false)
+   * @usage quant:run-queue --unlock
```
```suggestion
   * @option unlock
   *   Allow a queue run even if another run is in progress (default is false)
   * @usage quant:run-queue --unlock
```

kepol replied:
I indented and I removed the period only to be consistent with the previous parameter description (though all really should have periods)

kepol commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,29 @@ private function getDrushPath() {
+    // Bail if another run is in progress.
+    if (!$options['unlock']) {
+      if (file_exists($lockFilePath)) {
+        $this->output()->writeln("<info>Seeding bailed. Another seed run is in progress.</info>");
```
```suggestion
        $this->output()->writeln("<info>A seed is currently running and is locked which prevents another seed from running right now.</info>");
```

kepol commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,29 @@ private function getDrushPath() {
     $cmd = $drushPath . ' queue:run quant_seed_worker';
     $this->output()->writeln("<comment>Using drush binary at $drushPath. Override with \$DRUSH_PATH if required.</comment>");
 
+    // Bail if another run is in progress.
```
```suggestion
    // If another seed is running and is locked, don't run this one.
```

kepol commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,29 @@ private function getDrushPath() {
+    if (!$options['unlock']) {
+      if (file_exists($lockFilePath)) {
+        $this->output()->writeln("<info>Seeding bailed. Another seed run is in progress.</info>");
+        return;
+      } else {
+        // No lock currently present. Create new lock file.
```
```suggestion
        return;
      }
      else {
        // No lock file exists, so create one.
```

barbun committed:
Apply suggestions from code review

Great suggestions @kepol

Co-authored-by: Kristen Pol <github@kristen.org>

kepol commented:
I managed to get it into a weird state... I think because I used Ctrl-C in the middle of a run so that the lock file wasn't removed. Even though I ran --unlock afterwards. Not sure how to deal with this scenario though other than an option/command to have the lock file removed.

```
Kristens-MacBook-Pro:quantcdn kristenpol$ ddev drush quant-queue-run
Forking seed worker.
Using drush binary at /var/www/html/vendor/bin/drush. Override with $DRUSH_PATH if required.
 [success] Processed 0 items from the quant_seed_worker queue in 0.03 sec.
 [success] Processed 0 items from the quant_seed_worker queue in 0.02 sec.
 [success] Processed 0 items from the quant_seed_worker queue in 0.02 sec.
 [success] Processed 0 items from the quant_seed_worker queue in 0.02 sec.
 [success] Processed 0 items from the quant_seed_worker queue in 0.02 sec.
Seeding complete.
Kristens-MacBook-Pro:quantcdn kristenpol$ ddev drush quant-queue-run
Forking seed worker.
Using drush binary at /var/www/html/vendor/bin/drush. Override with $DRUSH_PATH if required.
A seed is currently running and is locked which prevents another seed from running right now.
Kristens-MacBook-Pro:quantcdn kristenpol$ ddev drush quant-queue-run
Forking seed worker.
Using drush binary at /var/www/html/vendor/bin/drush. Override with $DRUSH_PATH if required.
A seed is currently running and is locked which prevents another seed from running right now.
```

kepol commented:
Use cases:

1. Run locked and then unlocked at same time: both should run (lock file created by first run)
2. Run locked and then locked at same time: second shouldn't run (lock file created by first run)
3. Run unlocked and then unlocked at the same time: both should run (no lock file created)
4. Run unlocked and then locked at the same time: both should run (lock file created by second run)

kepol commented:
I wrote up the use cases I tested and they seem okay I think. The main issue is when doing the Ctrl-C when a lock file has been created and it leaves the lock file. You have to manually remove the lock file right now.

kepol commented:
Maybe @steveworley or @stooit have thoughts on how to handle ^

stooit commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -91,6 +107,11 @@ public function message($options = ['threads' => 5]) {
     }
 
+    // Remove lock file.
+    if ($options['unlock']) {
```
We would always want to remove the lockfile at this point I would think..

stooit commented:
There is no real clean way to catch ctrl+c that is cross-platform so would suggest adding a new drush command (e.g `drush quant:unlock-queue`) that simply removes the lockfile.

Then the use case here is:
* You are running `drush quant:run-queue` via cron and you expect it to consistently have more items than can be handled in the cron window
* You would want to have another cronjob that runs `drush quant:unlock-queue`) periodically (e.g 1x per day, depending on use case) to make sure the queue doesn't ever stay locked in case of an unexpected failure

kepol commented on src/Commands/QuantDrushCommands.php:
```diff
@@ -70,14 +70,30 @@ private function getDrushPath() {
    * @option threads
    *   Number of threads to use (default 5)
    * @usage quant:run-queue --threads=5
+   * @option unlock
+   *   Allow a queue run even if another run is in progress (default is false)
+   * @usage quant:run-queue --unlock
```
See Stu's comments about adding another drush command to be able to remove the lock file.

kepol commented:
@barbun Would you be able to add this additional command?

barbun commented:
> 

Hi @kepol sure, will do so a bit later. Thanks for your review :)

kepol commented:
Thanks @barbun... just checking in :)

Commit:
Added separate Drush command to unlock queue.

Commit:
Merge branch 'feature/bail-seed-run' of https://github.com/barbun/quantcdn-drupal10 into feature/bail-seed-run

barbun commented:
Hi @kepol apologies for the delays. I've now updated the code.

kepol commented:
Oh! Just noticing this. It missed this release but we can get it into the next one.

kepol commented:
This made it into the [1.5.3 release](https://www.drupal.org/project/quantcdn/releases/1.5.3).
